ios/ShareViewController/ShareViewController.swift

```diff
@@ -0,0 +1,210 @@
+        let group = DispatchGroup()
+        for attachment in attachments {
+            group.enter()
+            loadData(for: attachment, in: folder, group: group) {
+                error in
+                group.leave()
+                if let error = error {
+                    completion(error)
+                }
+            }
+        }
+    }
+
+    private func loadData(for attachment: NSItemProvider, in folder: URL, group: DispatchGroup, completion: @escaping (FileSaveError?) -> Void) {
+        let isURL = attachment.hasItemConformingToTypeIdentifier("public.url") && !attachment.hasItemConformingToTypeIdentifier("public.file-url")
+        let typeIdentifier = isURL ? (kUTTypeURL as String) : (kUTTypeData as String)
+        attachment.loadItem(forTypeIdentifier: typeIdentifier, options: nil) {
+            (data, error) in
+            DispatchQueue.main.async {
+                guard error == nil else {
+                    os_log("Sharing error: \(error!)")
+                    completion(.CouldNotLoad)
+                    return
+                }
+                if isURL, let url = data as? URL {
+                    self.handleURL(url, folder: folder, completion: completion)
+                }
+                else {
+                    self.handleData(data, folder: folder, completion: completion)
+                }
+            }
+        }
+    }
+
+    private func handleURL(_ url: URL, folder: URL, completion: @escaping (FileSaveError?) -> Void) {
+        let filename = "text_to_read.txt"
+        if let fileData = url.absoluteString.data(using: .utf8) as NSData? {
+            if let fileFinalPath = saveFileToFolder(folder: folder, filename: filename, fileData: fileData) {
+                completion(nil)
+            }
+            else {
+                completion(.CouldNotLoad)
+            }
+        }
+        else {
+            completion(.URLError)
+        }
+    }
+
+   private func handleData(_ data: Any?, folder: URL, completion: @escaping (FileSaveError?) -> Void) {
+        guard let data = data else {
+            os_log("Data is nil", type: .error)
+            completion(.CouldNotLoad)
+            return
+        }
+
+        if let dataString = data as? String {
+            handleStringData(dataString, folder: folder, completion: completion)
+        } else if let url = data as? NSURL {
+            handleURLData(url, folder: folder, completion: completion)
+        } else if let file = data as? UIImage {
+            handleImageData(file, folder: folder, completion: completion)
+        } else {
+            os_log("Received data of unhandled type", type: .error)
+            completion(.URLError)
+        }
+    }
+
+    private func handleStringData(_ dataString: String, folder: URL, completion: @escaping (FileSaveError?) -> Void) {
+        if !dataString.hasPrefix("file://") {
+            let filename = "text_to_read.txt"
```

Comment:
The `text_to_read.txt` appears three times, maybe we can extract it to a variable for maintanance. It's a minor one, though 😄